Bail early if path string is an empty string

phakeobj opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [X] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

...

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

Closes #17100

ret2libc commented on libr/main/rabin2.c:
```diff
@@ -893,6 +893,12 @@ R_API int r_main_rabin2(int argc, const char **argv) {
 	}
 	file = argv[opt.ind];
+
+	if (file && file[0] == '\0') {
```
why && and not || ? I think it should be || and in that case use R_STR_ISEMPTY

phakeobj replied:
because then it means that the if will handle the case where the user did not insert a file at all, where I would expect the program to print the print help and not complain that the path string is an empty string. the case of no file at all is handled in the if block right under my patch.

btw, I should probably call r_core_fini(&core) as in other places, before returning, noted.

phakeobj replied:
or maybe this _should_ be treated as when the user passes no file at all?

ret2libc replied:
Right, no then I guess it is fine as it is. And yes, you should probably call `r_core_fini` as you said. Thanks!

trufae commented on libr/main/rabin2.c:
```diff
@@ -893,6 +893,13 @@ R_API int r_main_rabin2(int argc, const char **argv) {
 	}
 	file = argv[opt.ind];
+
+	if (file && file[0] == '\0') {
```
Use R_STR_IS_EMPTY

ret2libc replied:
Please see discussion in https://github.com/radareorg/radare2/pull/17126/files#r443814345

trufae commented on libr/main/radare2.c:
```diff
@@ -690,6 +690,13 @@ R_API int r_main_radare2(int argc, const char **argv) {
 		}
 	}
+
+	if (pfile && pfile[0] == '\0') {
```
Same here

trufae replied:
I prefer !*pfile than pfile[0] == 0. Shorter snd concise. Can you change this in both places?

trufae commented on libr/main/rabin2.c:
```diff
@@ -893,6 +893,13 @@ R_API int r_main_rabin2(int argc, const char **argv) {
+
+	if (file && file[0] == '\0') {
+		eprintf ("Couldn't open empty path\n");
+		r_core_fini (&core);
```
Why return here and goto below? Would be better to have a single exit point

ret2libc replied:
The `goto` below is in another file though.

phakeobj replied:
I was trying to follow existing conventions in `r_main_radare2` and `r_main_rabin2`

trufae commented:
More issues to be fixed:

* r2 -i '' -
* r2 -P '' -
* r2 -R ''
* radiff2 '' ''

i dont think im missing any other, but please give it a try and see if you can find another confusing error/assert

Thanks!

phakeobj commented:
I will check it for every argument that takes a file path, in every tool in the suite. `radiff2` behavior, as an example, seems to handle empty paths without triggering any assert,

```
 ~ $ radiff2 '' ''
radiff2: Cannot open
```

Do you think I should add a test for empty string and not let it try to `opencore` the empty path strings?

trufae commented:
Yes, the earlier we check for invalid input filenames the faster we can spit out an error message. But this "Cannot open" message could be the same as the other cases

phakeobj commented:
I'm not satisfied with 4dfc203 but that's just WIP. I changed "couldn't" to "cannot" as it seems like this is the preferred tense in these errors in r2 suite. I feel bad about the code duplication, and I wonder if you have any r2 style (marco-ish maybe?) idea of how to make it less disturbing.

I've added the bails in rafind2 and rahash2 as well, didn't cover all the toolset yet, just wanted to get some feedback regarding to the direction of this issue, if this is going to where you intended.

ret2libc commented on libr/main/radare2.c:
```diff
@@ -592,6 +607,12 @@ R_API int r_main_radare2(int argc, const char **argv) {
 			r_config_set (r->config, "dbg.profile", opt.arg);
 			break;
 		case 'R':
+			file = opt.arg;
```
this looks wrong.
` -R [rr2rule] specify custom rarun2 directive`

`-R` is supposed to be used like `radare2 -R aslr=false -d /bin/ls`. It does not have a file as a argument. `-r` is probably what you wanted to check.

phakeobj replied:
truth is that now I'm not even sure if `-r` should be handled. The other two are OK as they're validating the input before appending it to the pre/post files list. I can't remember why I added `-R`, I think it's a leftover I forgot to delete.

phakeobj replied:
OK now I remember, and I think that the description, and the functionality of `-R` and `-r` is inverted. '-R' sets `customRarunProfile` which is later opened and appended to `dbg.profile` while, `-r` simply appends the given string to `dbg.profile`. Thoughts?

cc: @trufae

ret2libc replied:
So, `-R` creates a list of rarun2 directives, then it creates a temporary file, dump those entries in the temporary file and load dbg.profile with the temporary file.
`-r` directly gets a file and sets dbg.profile to that temporary file. So you should check whether the argument of `-r` is not an empty file.

phakeobj replied:
gotcha.
